src/Features/Core/Portable/PullMemberUp/CodeActionAndSolutionGenerator.cs

```diff
@@ -0,0 +1,197 @@
+            AnalysisResult result,
+            ICodeGenerationService codeGenerationService,
+            SyntaxNode targetNodeSyntax,
+            SolutionEditor solutionEditor,
+            Solution solution,
+            ImmutableDictionary<DocumentId, DocumentEditor> editorMap,
+            ImmutableDictionary<ISymbol, IEnumerable<SyntaxNode>> syntaxMap)
+        {
+            AddMembersToClass(result, codeGenerationService, targetNodeSyntax, solution, editorMap);
+            RemoveOriginalMembers(result, solution, editorMap, syntaxMap);
+            return solutionEditor.GetChangedSolution();
+        }
+
+        private void AddMembersToClass(
+            AnalysisResult result,
+            ICodeGenerationService codeGenerationService,
+            SyntaxNode targetNodeSyntax,
+            Solution solution,
+            ImmutableDictionary<DocumentId, DocumentEditor> editorMap)
+        {
+            var pullUpMembersSymbols = result._membersAnalysisResults.Select(memberResult => memberResult._member);
+
+            var options = new CodeGenerationOptions(reuseSyntax: true, generateMethodBodies: false);
+            var membersAddedNode = codeGenerationService.AddMembers(targetNodeSyntax, pullUpMembersSymbols, options: options);
+
+            var editor = editorMap[solution.GetDocumentId(targetNodeSyntax.SyntaxTree)];
+            editor.ReplaceNode(targetNodeSyntax, membersAddedNode);
+        }
+
+        private void RemoveOriginalMembers(
+            AnalysisResult result,
+            Solution solution,
+            ImmutableDictionary<DocumentId, DocumentEditor> editorMap,
+            ImmutableDictionary<ISymbol, IEnumerable<SyntaxNode>> syntaxMap)
+        {
+            foreach (var analysisResult in result._membersAnalysisResults)
+            {
+                foreach (var syntax in syntaxMap[analysisResult._member])
+                {
+                    RemoveDeclaration(editorMap[solution.GetDocumentId(syntax.SyntaxTree)], syntax);
+                }
+            }
+        }
+
+        private async Task<(ImmutableDictionary<DocumentId, DocumentEditor>, ImmutableDictionary<ISymbol, IEnumerable<SyntaxNode>>)> InitializeEditorMapsAndSyntaxMapAsync(
+            AnalysisResult result,
+            SyntaxNode targetSyntaxNode,
+            SolutionEditor solutionEditor,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            // Members and target may come from different documents,
+            // So EditorMap is used to save and group all the editors will be used.
+            var editorMapBuilder = ImmutableDictionary.CreateBuilder<DocumentId, DocumentEditor>();
+            // One member may have multiple syntaxNodes (e.g partial method).
+            // SyntaxMap is used to find the syntaxNodes need to be changed more easily.
+            var syntaxMapBuilder = ImmutableDictionary.CreateBuilder<ISymbol, IEnumerable<SyntaxNode>>();
```

Comment:
ok.  so the assumption here is that you're only going to edit documents containing the member decls and the target destination decls?  yes?  you don't have any other code that jumps elsewhere and tries to edit any other documents?

Reply:
Yes, I can also explain why I use DocumentId here rather than SyntaxTree. At the time when I write this, I used those two maps to solve the scenario when I need to do multiple changes on one SyntaxNode(e.g. , first add the members to destination and then change the destination to abstract). At that time I was thinking DocumentId is the only thing remain same during this process, so I use documentId here. But I just see your comments there is callback API for DocumentEditor.ReplaceNode, I can try to simplify the implementation here in next commit.